fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java

```diff
@@ -135,56 +162,76 @@ public static ScalarOperator translate(Expr expression, ExpressionMapping expres
 
     public static ScalarOperator translate(Expr expression) {
         IgnoreSlotVisitor visitor = new IgnoreSlotVisitor();
-        return visitor.visit(expression, null);
+        return visitor.visit(expression, new Context());
     }
 
-    public static ScalarOperator translateWithoutRewrite(Expr expression, ExpressionMapping expressionMapping,
-                                                         ColumnRefFactory columnRefFactory) {
-        ColumnRefOperator columnRefOperator = expressionMapping.get(expression);
-        if (columnRefOperator != null) {
-            return columnRefOperator;
-        }
+    private static final class Context {
 
-        Visitor visitor = new Visitor(expressionMapping, new ArrayList<>(), columnRefFactory);
-        ScalarOperator result = visitor.visit(expression, null);
+        public final boolean useSemiAnti;
+        public final List<Expr> outerExprs;
 
-        ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
-        result = scalarRewriter.rewrite(result, ScalarOperatorRewriter.DEFAULT_TYPE_CAST_RULE);
+        public Context() {
+            useSemiAnti = false;
+            outerExprs = Collections.emptyList();
+        }
 
-        requireNonNull(result, "translated expression is null");
-        return result;
+        public Context(boolean useSemiAnti, List<Expr> outerExprs) {
+            this.useSemiAnti = useSemiAnti;
+            this.outerExprs = outerExprs;
+        }
+
+        public Context clone(Expr node) {
+            List<Expr> outerExprs = Collections.emptyList();
+            if (node.getChildren().stream().filter(Subquery.class::isInstance).count() == 1) {
+                outerExprs = node.getChildren().stream().filter(c -> !(c instanceof Subquery))
+                        .collect(Collectors.toList());
+            }
+            if (!this.useSemiAnti && outerExprs.isEmpty()) {
+                return this;
+            }
+            return new Context(false, outerExprs);
+        }
```

Comment:
emmm，we need clone context everywhere？
origin logic:  
1. only (top expression) will set `useSemiAnti = true`,  other place must be false
2. and expression will send parent's useSemiAnti to children, or/not must be false

and outerExprs is for send outer's columnRef, like 
```
select * from xx where x1 = (select .....)
```
then the outerExprs is `x1`.
 
OK,  it's not problem, just help you understand subquery

Reply:
> and expression will send parent's useSemiAnti to children, or/not must be false

Yes, only compound `and` should send parent's useSemiAnti to children, and the others should be false.